### PR TITLE
[integrationtests] A way of testing rules within openHAB in isolation

### PR DESCRIPTION
**Pull request by FlorianSW**
openHAB is quite a powerful tool for smart home automation. It serves a lot of user needs, both in UI and textual configuration, and is quite easily extensible by using existing addons or developing own addons. The rule engine allows a user to easily do quite everything to automate whatever needs to be automated.

One downside when _using_ openHAB is, that the setup can become quite compley over the time, especially with a lot of rules, items and things in place. This makes it pretty hard to make changes to the setup and afterwards stand there and say, that anything is working like expected. A big problem here is, that testing your setup mostly requires you to have different actual devices up and running, with which you want to integrate with (e.g. knx, Hue or zigbee devices to name just a few).

However, testing the main logic of the setup, which mainly consists of items that are automated with rules, can be done pretty easily without having the need to actually control devices. Having a way to test the rule logic would at least be a step in the right direction to make the setup of openHAB more reliable when changing and configuring it further.

As there was no way of testing rules in an automated fashion (the normal way would be to trigger a rule in your production setup and see if it is doing what you expect it to do), I took some time to find a way to make this easier, more isolated and being able to run the tests on a separate, non-production environment. What came out of it is the code in this PR. Let me explain what the idea behind it is:
* The user should be able to write tests in java
-> I think it is no beneficial to use xtext for that or even introduce another intermediate programming language, as the syntax of java (at least in the way the tests are designed) should be pretty easy to adopt from the rules syntax)
* The first step allows users of the textual configuration to tests this one
-> it's easier to move the textual configuration (or even "just" point to it) instead of the saved UI configuration, however, the design of the tests should not require the textual configuration, so that a UI saved configuration _should_ work fine, too, is, however, not the goal.
* the tests run in isolation
-> Instead of having to use the openHAB setup of your home, the tests should run in an own, separate and isolated openHAB installation.
* the tests are independent of the bindings and other addons that serve the items
-> As the goal of the tests that can be written with this introduced framework is to test the logic of your rules, the tests are run _without_ addons installed and even without taking the things into account. That means, that the focus of the tests is the interaction with items and the expected outcome of items (which is the logic of the rule)
* the tests should be able to run in an automated way
-> Automated here also means in an "non-interactive" way. It's ok to run the tests in eclipse with a run configuration, however, for a good integration strategy, one might want to automatically test their rules with a proper CI system, like Travis, Jenkins or Concourse. This requires the tests to be able to be run with a set of command line tools, e.g. mvn integration-test

### CODE REVIEW

**Comment by FlorianSW** [comment]
While all of the above points are already working with the code in this PR (see the ExampleTest.java file for a pretty easy example test, which should run with the sample rules and items of the openhab-distro repo), my pain point is the automated, non-interactive run. So, this PR is not only a request for feedback about the idea and the implementation, but also a request for help, as I'm totally not an OSGi or Eclipse Plugin/tycho expert at all :( When I run the `mvn integration-test` command, there're two observations:
* On a system without the eclipse-smarthome libraries in the .m2 repository, I get the error message, that I requested the "osgi bundle org.eclipse.smarthome.automation.module.script.rulesupport 0.0.0" could not be resolved. I don't know how to fix that, as the openHAB target platform contains this bundle.
* On a system with the eclipse-smarthome libraries in the .m2 repository (no idea when they were installed, though) I get this error message: `java.lang.RuntimeException: Cannot create a resource for 'http://www.eclipse.org/2008/Xtext'; a registered resource factory is needed` (see also [this question](https://community.openhab.org/t/maven-build-openhab-plugin-cannot-create-a-resource-for-http-www-eclipse-org-2008-xtext-a-registered-resource-factory-is-needed/64000) in the forum, where I did not get any answer, yet).

So, if anyone more familar with the build configuration and system has an idea, I would be glad for any help :)

**Comment by maggu2810** [comment]
Hi @FlorianSW, perhaps you already realized that things changed a lot.
openHAB Core is no the framework / core system used by openHAB itself -- but also from other products.
We dropped the usage of Tycho etc.

**Comment by FlorianSW** [comment]
@maggu2810 thanks for the comment and please excuse my late reply :)

> openHAB Core is no the framework / core system used by openHAB itself -- but also from other products.

I kind of know that, however, the target features, which this PR targets to allow to test, are implemented in this repo, that's why I think it's a good place to integrate this one into :) No matter, if the users will be in the openHAB product or any other product (which will most likely provide the same automation features, and if not, the ruletests bundle does not need to be integrated there, either :)).

> We dropped the usage of Tycho etc.

Saw that already, yes, and from my personal point of view, that is a step into the right direction I think :) Nontheless, with rebasing my work ontop of the mavenized openhab-core also makes it possible to run the tests from the command line with a simple maven command :)

**Comment by maggu2810** [comment]
> I kind of know that, however, the target features, which this PR targets to allow to test, are implemented in this repo, that's why I think it's a good place to integrate this one into

I did not complain about the repository your PR has been created for, just want to highlight that change.

**Comment by FlorianSW** [comment]
Oh, sorry if, because of my comment, you got the feeling I thought you complained about something. That was not the case, I just wanted to explain some of my thoughts. So, sorry for that :)

**Comment by FlorianSW** [comment]
It would be great to get some opinions about, at least, the direction of this PR. I think it's valuable for people who want to ensure, that their home automation logic they built is doing what they expect it to do whenever they change something. The essential point of having this confidence are automated tests, and testing rules is something which can be done in isolation (at least in a better way as tests which really interact with actual readl devices).

So, I really would like to hear some opinions here :)

File: bundles/org.openhab.core.karaf/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/bin/
```
Comment (maggu2810) [inline bundles/org.openhab.core.karaf/.gitignore]: Should not be part of this PR.

File: bundles/org.openhab.core.rulestest/.classpath [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
```
Comment (maggu2810) [inline bundles/org.openhab.core.rulestest/.classpath]: No PDE anymore.

File: bundles/org.openhab.core.rulestest/.project [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.pde.SchemaBuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.pde.PluginNature</nature>
```
Comment (maggu2810) [inline bundles/org.openhab.core.rulestest/.project]: No PDE anymore.

File: bundles/org.openhab.core.rulestest/pom.xml [hunk 4]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (maggu2810) [inline bundles/org.openhab.core.rulestest/pom.xml]: Can you please use two spaces for indentation?

We can fix this later by using an automatic task (would like to do it for the other ones, too).

File: itests/org.openhab.core.rulestest.tests/.classpath [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
```
Comment (maggu2810) [inline itests/org.openhab.core.rulestest.tests/.classpath]: No PDE anymore.

File: itests/org.openhab.core.rulestest.tests/.project [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.pde.SchemaBuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.pde.PluginNature</nature>
```
Comment (maggu2810) [inline itests/org.openhab.core.rulestest.tests/.project]: No PDE anymore.

File: bundles/org.openhab.core.karaf/.gitignore (hunk 1 above)
Comment (FlorianSW) [reply bundles/org.openhab.core.karaf/.gitignore]: Yep, don't know how this ended up here :)

File: bundles/org.openhab.core.rulestest/.classpath (hunk 2 above)
Comment (FlorianSW) [reply bundles/org.openhab.core.rulestest/.classpath]: +1

File: bundles/org.openhab.core.rulestest/.project (hunk 3 above)
Comment (FlorianSW) [reply bundles/org.openhab.core.rulestest/.project]: +1

File: itests/org.openhab.core.rulestest.tests/.classpath (hunk 5 above)
Comment (FlorianSW) [reply itests/org.openhab.core.rulestest.tests/.classpath]: +1

File: itests/org.openhab.core.rulestest.tests/.project (hunk 6 above)
Comment (FlorianSW) [reply itests/org.openhab.core.rulestest.tests/.project]: +1

File: bundles/org.openhab.core.rulestest/pom.xml (hunk 4 above)
Comment (FlorianSW) [reply bundles/org.openhab.core.rulestest/pom.xml]: ok

**Comment by kaikreuzer** [comment]
@maggu2810 As you had already reviewed this PR - could you have a look at the updates that were done meanwhile?

**Comment by maggu2810** [comment]
@kaikreuzer @FlorianSW I cannot have a look at the next two weeks, but after that I assume there will be some spare time to look into.

@FlorianSW Do you think it would make sense to do some chatting then so you can tell me what to do and assist in my look into?

**Comment by FlorianSW** [comment]
> @kaikreuzer @FlorianSW I cannot have a look at the next two weeks, but after that I assume there will be some spare time to look into.
> 
> @FlorianSW Do you think it would make sense to do some chatting then so you can tell me what to do and assist in my look into?

If this helps you, of course. However, I'm a bit busy, so I'm not sure when I've time. Additionally, I haven't setup openHAB-core on my new development computer at all (still struggling with the setup of the openhab2-addons) :( However, I can give insights in what my thoughts were and what is needed to work with this PR :)

**Comment by wborn** [comment]
Hello @FlorianSW, are you still around and what is the status of this PR?
Looks like this PR will need many updates before it will build again.

**Comment by FlorianSW** [comment]
> Hello @FlorianSW, are you still around and what is the status of this PR?
> Looks like this PR will need many updates before it will build again.

I'm not thaaat active in openHAB anymore tbh. I don't think I'll have much time to put into this PR to make it work again. And given the massive amount of changes to openhab-core since I last worked on this PR, I'm even not sure if this is still the best approach. I would therefore abandon this PR. It will still be around if someone will have time to come back to it, if it still contains valid code snippets to take over :)

**Comment by wborn** [comment]
Thanks for the update! :+1: Even if PRs don't make it, you always learn something while creating them. :slightly_smiling_face:
